toolchain/sem_ir/class.cpp

```diff
@@ -10,6 +10,33 @@
 
 namespace Carbon::SemIR {
 
+auto Class::GetAdaptedType(const File& file, SpecificId specific_id) const
+    -> TypeId {
+  if (!adapt_id.is_valid()) {
+    return TypeId::Invalid;
+  }
+  if (base_id == SemIR::InstId::BuiltinErrorInst) {
+    return TypeId::Error;
+  }
+  return TypeId::ForTypeConstant(GetConstantValueInSpecific(
+      file, specific_id,
+      file.insts().GetAs<AdaptDecl>(adapt_id).adapted_type_inst_id));
+}
+
+auto Class::GetBaseType(const File& file, SpecificId specific_id) const
+    -> TypeId {
+  if (!base_id.is_valid()) {
+    return TypeId::Invalid;
+  }
+  if (base_id == SemIR::InstId::BuiltinErrorInst) {
+    return TypeId::Error;
+  }
+  CARBON_CHECK(base_id.index >= 0);
+  return TypeId::ForTypeConstant(GetConstantValueInSpecific(
+      file, specific_id,
+      file.insts().GetAs<BaseDecl>(base_id).base_type_inst_id));
+}
```

Comment:
Looking at the rename of `adapted_type_inst_id`, had you considered renaming `base_type_inst_id` and `adapted_type_inst_id` to something like `extended_type_inst_id`? It looks like adding `AnyExtendedDecl` (or a template on `AdaptDecl`/`BaseDecl`) would allow code sharing here:

```suggestion
// Returns the extended type for an `adapt` or `base` declaration.
static auto GetExtendedType(const File& file, SpecificId specific_id, InstId extended_id) -> TypeId {
  if (!extended_id.is_valid()) {
    return TypeId::Invalid;
  }
  if (extended_id == SemIR::InstId::BuiltinErrorInst) {
    return TypeId::Error;
  }
  return TypeId::ForTypeConstant(GetConstantValueInSpecific(
      file, specific_id,
      file.insts().GetAs<AnyExtendedDecl>(extended_id).extended_type_inst_id));
}

auto Class::GetAdaptedType(const File& file, SpecificId specific_id) const
    -> TypeId {
  return GetExtendedType(file, specific_id, adapt_id);
}

auto Class::GetBaseType(const File& file, SpecificId specific_id) const
    -> TypeId {
  return GetExtendedType(file, specific_id, base_id);
}

Reply:
Seems like a nice idea. The name doesn't work for me, though -- an adapted type need not be extended (you can use `adapt` without `extend`), and I think there's a reasonable chance we'll decide at some point we want to allow `base` without `extend` too. Will take to #naming.

Reply:
Done, went with "foundation declaration" for base type or adapted type.